src/Tools/dotnet-monitor/dotnet-monitor.csproj

```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```

Comment:
The downside to this change is that dotnet-monitor now sees itself when querying for processes. This shouldn't be an issue when running in the Docker container that we are creating because we set COMPLUS_EnableDiagnostics to 0. However, if you run the tool as-is without setting that environment variable, then it'll see itself. Although, it being able to perform diagnostics on itself doesn't seem to be an issue.

I'm not certain that we can accurately filter itself when discovering processes.
- When it's running in the 3.1 runtime, it only has the ability to publish its diagnostic port at the default location. We might be able to just check the process ID and filter on that, however, that would break the current usage in the Kubernetes setup where the /tmp directory for both the application container and dotnet-monitor container are mounted to the same volume.
- When it's running in the 5.0 runtime (roll forward), it still has the same issues as the 3.1 runtime scenario when the diagnostic port is in listen mode. Additionally, if its diagnostic port is configured to connect (because it happens to inherit environment variables), then it'll be able to see itself and we cannot filter on process ID (because there can be collisions). I don't think that the managed code would be able to get the runtime instance ID without doing something like connecting to its own diagnostic port (because the default port will be in listen mode), issue the GetProcessInfo command, and then filter the usual process list so that the matching one is excluded. @josalem, with your changes to allow multiple diagnostic ports to be specified in .NET 5, would all of the ports report the same runtime identifier when issuing GetProcessInfo or getting in from the advertise data?

Alternatively, I think if runtime could look in the runtime configuration (e.g. dotnet-monitor.runtimeconfig.json) for a setting that would describe whether diagnostics should be enabled for the process or not so that we can disable it as published rather than requiring that an environment variable is set, then that would solve this problem and still allow the user to clear that setting if they really wanted to perform diagnostics on the dotnet-monitor process.

@wiktork, @noahfalk, @josalem do you have any thoughts on this?

Reply:
I think a good starting point:
For multi-container scenarios, I think it's reasonable to rely on the environment variable.
For single container scenarios, it's possible to spawn a child process with a different environment variable than the parent.
On Windows, you should have an explicit check against the pid to avoid self-diagnosis.

Reply:
I think we can simplify it to saying that the current process is only filtered from the list of processes when not running in a container and querying the list of processes in forward / connect mode.

Reply:
I haven’t had a chance to think through your other questions, but I will ponder them over the weekend.  As for the runtime identifier: it  would be the same for all communication from a given runtime instance regardless of the port it goes across.

Reply:
> COMPLUS_EnableDiagnostics to 0

I think of this as an option of last resort. At minimum I'd hope to say that for 5.0 onwards we shouldn't need to disable diagnostics for any 'typical scenario'. I think 'typical' means:
- listen mode ports on the same machine/container (no possibility of PID conflict, monitor filter itself by PID)
- connect mode ports anywhere (either the monitor shouldn't connect to itself, or it should filter itself by runtime id)

Reply:
I've moved the filter from ClientEndpointInfoSource to DiagnosticServices. It now self filters on process ID if dotnet-monitor is connecting to runtime instances and filters on runtime instance ID if dotnet-monitor is listening to runtime instances. 

Although, this still does break the mounted /tmp scenario for .NET Core 3.1 containers. I think an appropriate fix for this is to have it check if diagnostics are enabled for dotnet-monitor, then it can filter itself; if disabled, do not filter itself.

Reply:
I've changed it to filter itself only when diagnostics are enabled on itself.